connect/src/com/telenor/connect/headerenrichment/HeLogic.java

```diff
@@ -0,0 +1,210 @@
+        boolean authenticateNow = finishedUnSuccessfully || promptBlocksUseOfHe || cantDirectNetworkTraffic;
+        if (authenticateNow) {
+            callCallbacks(showLoadingCallback, heTokenCallback);
+            return;
+        }
+
+        if (isHeTokenRequestOngoing) {
+            setFutureHeTokenCallback(parameters, showLoadingCallback, heTokenCallback, logSessionId, useStaging);
+            return;
+        }
+
+        boolean heWasNeverInitialized = heTokenSuccess && heTokenResponse == null;
+        boolean tokenIsExpired = heTokenResponse != null && new Date().after(heTokenResponse.getExpiration());
+        if (heWasNeverInitialized || tokenIsExpired) {
+            setFutureHeTokenCallback(parameters, showLoadingCallback, heTokenCallback, logSessionId, useStaging);
+            initializeHeaderEnrichment(useStaging, logSessionId);
+            return;
+        }
+
+        callCallbacks(showLoadingCallback, heTokenCallback);
+    }
+
+    private static void callCallbacks(ShowLoadingCallback showLoadingCallback, HeTokenCallback heTokenCallback) {
+        if (showLoadingCallback != null) {
+            showLoadingCallback.stop();
+        }
+        heTokenCallback.done();
+    }
+
+    private static void setFutureHeTokenCallback(
+            final Map<String, String> parameters,
+            final ShowLoadingCallback showLoadingCallback,
+            final HeTokenCallback heTokenCallback,
+            final String logSessionId,
+            final boolean useStaging) {
+        HeLogic.heTokenCallback = new HeTokenCallback() {
+            @Override
+            public void done() {
+                HeLogic.heTokenCallback = null;
+                handleHeToken(parameters, showLoadingCallback, heTokenCallback, logSessionId, useStaging);
+            }
+        };
+    }
+
+    private static void handleHeTokenResult(HeTokenResponse heTokenResponse) {
+        isHeTokenRequestOngoing = false;
+        HeLogic.heTokenResponse = heTokenResponse;
+        heTokenSuccess = heTokenResponse != null;
+        if (heTokenCallback != null) {
+            heTokenCallback.done();
+        }
+    }
+
+    public static boolean failedToGetToken() {
+        return !heTokenSuccess || heTokenResponse == null;
+    }
+
+    public static HeTokenResponse getHeTokenResponse() {
+        return heTokenResponse;
+    }
+
+    public static boolean isCellularDataNetworkConnected() {
+        if (connectivityManager == null) {
+            return false;
+        }
+        NetworkInfo networkInfo;
+        if (cantDirectNetworkTraffic) {
+            networkInfo = connectivityManager.getNetworkInfo(ConnectivityManager.TYPE_MOBILE);
+        } else {
+            if (cellularNetwork == null) {
+                return false;
+            }
+            networkInfo = connectivityManager.getNetworkInfo(cellularNetwork);
+        }
+        return (networkInfo != null) && networkInfo.isConnected();
```

Comment:
You have parentheses around the first conditional here, and this differs from elsewhere (see f ex line 198). Is there a reason for the parentheses here?

Reply:
Good question - I don't know :p (not the original author). Looking at the original commit I don't see any reason for it: 04be260611697b0c3f96d1797b40778aaa2102e1

Reply:
d322499